rfc/rfc-8/rfc-8.md

```diff
@@ -0,0 +1,209 @@
+on the number of records in the dataset (available after Step 2 above). This should simplify rollout as the user does not
+have to worry about the number of fileGroups for optimal performance. Configs will allow specifying the number of fileGroups
+too.
+
+
+### Metadata Index lookup:
+
+For the incoming upsert records, given their keys, tag their current location. The key lookup would require the following steps:
+
+1. Generate the list of keys to be looked up (extract HoodieKeys from the upsert records)
+2. Lookup all the keys from the HUDI Metadata Table
+   1. Keys are partitioned based on the hash as the HUDI Metadata Table mappings are saved in various fileGroups (count fixed at initialization time) with each fileGroup saving a portion of the key space
+   2. Each partition of keys is looked up in parallel from the fileGroup using various Executors
+3. Tag the location, where a mapping was found in HUDI Metadata Table, back to upsert records 
+
+Given N fileGroups in the record index, an indexing lookup of M keys is reduced to N lookups of M/N keys in parallel. Hence, 
+for fastest lookup operation, the number of executors for the writer process should be >= N.
+
+This also means that lookup from record index can be scaled with growing data size by:
+1. Increasing the number of fileGroups (N in the above example)
+2. Using at least N or greater executors for the indexing process
+
+HDFS based experiments have shown than on average key lookups from HFile in HUDI Metadata Table complete in 1-2msec. 
+So for lookup of M keys we expect ballpark time of K + M / N * 2msec where K is the overhead of opening HFile (~100msec) 
+and merging the log files. Periodic compaction of Metadata Table keeps the value of K lower.
+
+
+## Implementation
+1. No changes to the HoodieIndex public interface.
+2. A new index type will be added - RECORD_LEVEL
+
+
+### Writer flow:
+Let's walk through the writer flow to update the record index.
+
+Whenever a new commit is getting applied to metadata table, we do the following.<br>
+1. Parse the WriteStatus to determine the record which have been inserted into the dataset
+   1. Such records have new location (HoodieRecord::getNewLocation()) but no current location ((HoodieRecord::getCurrentLocation()))
+2. Create new records for each record keys being added to the dataset
+3. Commit all these records to metadata table.
+
+We need to ensure that WriteStatus tracks all written records keys for every commit.
+
+
+### Reader flow:
+When a new batch of write is ingested into Hudi, we need to tag the records with their
+original file group location. Refer to Metadata Index lookup section for more details.
+
+
+## Rollout/Adoption Plan
+* Record Index will be available in 0.14.0 release
+* Metadata Table scheme will be upgraded as part of release upgrade process
+* Record Index will be disabled by default and can be enabled by setting the write configs 
+
```

Comment:
Can you add some lines on known limitations or future work? Otherwise lgtm.

Reply:
Added